core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2231,11 +2233,20 @@ object ZIO {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    val size      = as.size
+    val resultArr = new AtomicReferenceArray[B](size)
+
+    val wrappedFn: ZIOFn1[(A, Int), ZIO[R, E, Any]] = ZIOFn(fn) {
+      case (a, i) => fn(a).tap(b => ZIO.effectTotal(resultArr.set(i, b)))
+    }
+
+    foreachPar_(as.toStream.zipWithIndex)(wrappedFn).as(
```

Comment:
Do we need the `toStream` call here?

Reply:
This is in response to a review by @vasilmkd way back up at the top with regards to minimizing traversals of the iterable. I originally resolved it by doing a foldLeft, however after reimplementing foreachPar in terms of foreachPar_ this was no longer possible so I replaced it by converting the interable to a stream (the stream just wraps an iterator so isn't especially expensive). I'm not opposed to getting rid of it.

Reply:
Got it. The cost of any of this is probably dwarfed by everything else we're doing so fine either way. We could use `iterator` here potentially.